### PR TITLE
firefox: add default containers

### PR DESCRIPTION
**Pull request by rhoriguchi**
### Description

`userContextIdInternal.webextStorageLocal` is used by certain extensions without they don't work correctly <https://searchfox.org/mozilla-central/rev/762f24e00a9548d80ebba1b985c871ba6d9b829d/toolkit/components/contextualidentity/ContextualIdentityService.sys.mjs#93-96>. `userContextIdInternal.thumbnail` is some default added by firefox, so I just added it as well.

> This userContextId is used by ExtensionStorageIDB.jsm to create an IndexedDB database
> opened with the extension principal but not directly accessible to the extension code
> (do not change the userContextId assigned here, otherwise the installed extensions will
> not be able to access the data previously stored with the browser.storage.local API).

### Checklist

- [X] Change is backwards compatible.

- [X] Code formatted with `./format`.

- [X] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [X] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [X] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

- @DamienCassou 
- @rycee

### CODE REVIEW

File: modules/programs/firefox.nix [hunk 1]
```diff
@@ -663,6 +680,19 @@ in {
...
+      (let
+        getContainers = profiles:
+          flatten (mapAttrsToList (_: value: value.containers) profiles);
+
+        findInvalidContainerIds = profiles:
+          filter (container: container ? "id" && container.id < 4294967294)
+          (getContainers profiles);
+      in {
+        assertion = cfg.profiles == { }
+          || length (findInvalidContainerIds cfg.profiles) != 0;
```
Comment (DamienCassou) [inline modules/programs/firefox.nix:692]: I think `!= 0` should be changed to `== 0`.

File: modules/programs/firefox.nix [hunk 2]
```diff
@@ -663,6 +680,19 @@ in {
           (", namely " + concatStringsSep ", " defaults);
       })
 
+      (let
+        getContainers = profiles:
+          flatten (mapAttrsToList (_: value: value.containers) profiles);
```
Comment (DamienCassou) [inline modules/programs/firefox.nix:685]: given that `value.containers` is an attribute set, `mapAttrsToList` is going to return a list of attribute sets. There is nothing to flatten here. For example, the function will probably return something like:

```
[
  { "shopping" = ...; } // 1st profile
  { "shopping" = ...; } // 2nd profile
]
```

I don't think this is what the rest of the code expects.

**Review by DamienCassou — CHANGES_REQUESTED** [review]
Thank you for this. I think the code is broken but the idea is worth implementing. Can you please double-check?

File: modules/programs/firefox.nix (hunk 1 above)
Comment (rhoriguchi) [reply modules/programs/firefox.nix]: You are right but also had to switch the filter function.

File: modules/programs/firefox.nix (hunk 2 above)
Comment (rhoriguchi) [reply modules/programs/firefox.nix]: Yeah take a look at the new `getContainers` function, I've also added `attrValues` which then requires `flatten`. Please take an other look.

**Review by DamienCassou — APPROVED** [review]
This is great work, thank you!

**Comment by nyabinary** [comment]
Should be good to merge right?

**Comment by rhoriguchi** [comment]
> Should be good to merge right?

Yes, is ready to be merged.

**Comment by rycee** [comment]
Thanks! Merged to master now https://github.com/nix-community/home-manager/pull/4851

**Comment by brckd** [comment]
`profile-setting-expected-containers.json` needs to be updated

**Pull request merged**